polling_stations/apps/data_importers/data_types.py

```diff
@@ -0,0 +1,134 @@
+"""
+Data type classes used by base importers
+"""
+
+import abc
+import logging
+
+from addressbase.models import get_uprn_hash_table, UprnToCouncil
+from uk_geo_utils.helpers import Postcode
+
+
+class CustomSet(metaclass=abc.ABCMeta):
+    def __init__(self):
+        self.elements = set()
+
+    def add(self, element):
+        self.elements.add(self.build_namedtuple(element))
+
+    @abc.abstractmethod
+    def build_namedtuple(self, element):
+        pass
+
+
+class AssignPollingStationsMixin(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def get_polling_station_lookup(self):
+        pass
+
+    @property
+    def council_id(self):  # TODO Deal with old_to_new council_ids map
+        for e in self.elements:
+            if isinstance(e, dict):
+                return e["council"].council_id
+            else:
+                return e.council.council_id
+
+    def update_uprn_to_council_model(self, polling_station_lookup=None):
+        if not polling_station_lookup:
+            polling_station_lookup = self.get_polling_station_lookup()
+
+        uprns_in_council = UprnToCouncil.objects.filter(lad=self.council_id)
+
+        for polling_station_id, uprns in polling_station_lookup.items():
+            uprns_in_council.filter(uprn__in=uprns).update(
+                polling_station_id=polling_station_id
+            )
+
+
+class StationSet(CustomSet, AssignPollingStationsMixin):
+    NotImplemented
+
+
+class DistrictSet(CustomSet, AssignPollingStationsMixin):
+    NotImplemented
+
+
+class AddressList(AssignPollingStationsMixin):
+    def __init__(self, logger):
+        self.elements = []
+        self.logger = logger
+
+    def append(self, address):
+
+        if (
+            not address["address"]
+            or not address["postcode"]
+            or not address["uprn"]
+            or not address["council"]
+            or not address["polling_station_id"]
+        ):
+            self.logger.log_message(
+                logging.DEBUG,
+                "Record with empty required fields found:\n%s",
+                variable=address,
+                pretty=True,
+            )
+            return
+
+        self.elements.append(address)
+
+    def get_uprn_lookup(self):
+        # for each address, build a lookup of uprn -> set of station ids
+        uprn_lookup = {}
+        for record in self.elements:
+            uprn = record["uprn"]
+            if not uprn:
+                continue
+            if uprn in uprn_lookup:
+                uprn_lookup[uprn].add(record["polling_station_id"])
+            else:
+                uprn_lookup[uprn] = {record["polling_station_id"]}
+
+        return uprn_lookup
+
+    # TODO be more clever to report on duplicates.
```

Comment:
Just having a look over this module, it looks like one of the big things missing now is logging. In some ways that's good because a lot of the stuff we were logging is now not an issue but this is a good example of the sort of thing where it probably is useful to have a bit more visibility when you're trying to track down an issue. Step one on the road to doing something smarter is probably to log the dupes.

> check this sort of thing* when a csv is uploaded

I suspect you probably don't want to fail a file at the upload stage because of UPRN dupes. Anecdotally, you will see very few input files that contain zero UPRN dupes. Many of these will indicate errors, but occasionally they are also semi-legit (e.g: in Addressbase you've got '1 High Street' and then in the council's file you've got 'Top Flat 1 High Street' and 'Bottom Flat 1 High Street' both with the UPRN attached to '1 High Street' in AddressBase type stuff). Now we can just discard individual bad records with no worries, I agree just binning all dupe UPRN records without attempting to unpick why or do anything clever on import is the right thing to do :+1: It is certainly  in the spirit of trying to reduce manual overhead and scale the service.